### PR TITLE
Add datalab notebook

### PR DESCRIPTION
**Pull request by elisno**
This notebook focuses on running Datalab with a subset of the Caltech-256 dataset. It uses precomputed pred_probs and features.

Remaining todos:

- [x] Move features and pred_probs to S3.
- [x] Add notebook that generates features + pred_probs and saves them. See https://github.com/cleanlab/examples/pull/50#issuecomment-1513548981.

### CODE REVIEW

**Comment by jwmueller** [comment]
Can you add a second notebook that shows how to generate the pred_probs and features + save them to the files that are then loaded?

**Comment by elisno** [comment]
> Can you add a second notebook that shows how to generate the pred_probs and features + save them to the files that are then loaded?

Ah yes, thanks! Added todo in the description.

File: datalab/requirements-datalab.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+git+https://github.com/elisno/cleanlab.git@datalab#egg=cleanlab[datalab]
+matplotlib
+ipykernel
```
Comment (jwmueller) [inline datalab/requirements-datalab.txt]: why is ipykernel required?

Comment (elisno) [reply datalab/requirements-datalab.txt]: Just for local environment. Removed.

**Comment by huiwengoh** [comment]
[nit] I think we can remove the `_subset` suffix from the titles of the notebook, it feels a little distracting imo. 
The main one can just be `datalab.ipynb`, perhaps the other can be something like `train_caltech.ipynb`?

File: datalab/train_subset.ipynb [hunk 2]
```diff
@@ -0,0 +1,501 @@
+{
```
Comment (jwmueller) [inline datalab/train_subset.ipynb]: Rename this notebook: train_image_classifier.ipynb

File: datalab/train_subset.ipynb [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+    "\n",
+    "from PIL import Image"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Dataset Preparation\n",
```
Comment (jwmueller) [inline datalab/train_subset.ipynb]: If it's easy to do, I'd skip this entire data prep section and just load the prepped dataset from:

```
!wget -nc 'https://cleanlab-public.s3.amazonaws.com/Datalab/caltech256-subset.tar.gz'
```

This dataset prep section doesn't really help somebody better understand how to train an image classifier on their own data which is the purpose of this notebook

Comment (jwmueller) [reply datalab/train_subset.ipynb]: If not easy to change though, then just keep as is.

File: datalab/train_subset.ipynb [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Feature Extraction\n",
+    "\n",
+    "After training the model, we will compute predicted probabilities for the entire dataset using the trained models from each fold.\n",
```
Comment (jwmueller) [inline datalab/train_subset.ipynb]: why are you discussing predicted probabilities in this section?

File: datalab/train_subset.ipynb [hunk 5]
```diff
@@ -0,0 +1,501 @@
...
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Image Classifier Training on Caltech-256 Subset\n",
+    "\n",
+    "In this notebook, we will train a Swin Transformer model for image classifier on a subset of the Caltech-256 dataset using the Timm library and Pytorch. The main steps include dataset sampling and precrocessing, model training with k-fold cross-validation, and feature extraction.\n"
+   ]
+  },
```
Comment (jwmueller) [inline datalab/train_subset.ipynb]: clarify somewhere that Datalab does not actually require both features and pred_probs, but providing both allows it to more comprehensively detect issues in a dataset.

File: datalab/train_subset.ipynb [hunk 6]
```diff
@@ -0,0 +1,501 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### K-Fold Cross-Validation and Training\n",
+    "\n",
+    "In this notebook, we use k-fold cross-validation to train the model and extract out-of-sample predicted probabilities for all data points.\n",
+    "\n",
+    "During the training process, we set aside a validation split on a per-fold basis to allow early stopping for each fold. This approach helps us prevent overfitting and obtain a better estimate of the model's performance.\n",
+    "\n",
+    "While we're not specifically interested in the model artifacts themselves, we aim to get a general idea of whether the chosen model architecture is accurate enough for our purpose."
```
Comment (jwmueller) [inline datalab/train_subset.ipynb]: warn that this cell may take a long time to execute and should be run with a GPU

File: datalab/train_subset.ipynb (hunk 3 above)
Comment (elisno) [reply datalab/train_subset.ipynb]: Easy enough.

File: datalab/train_subset.ipynb (hunk 4 above)
Comment (elisno) [reply datalab/train_subset.ipynb]: Renamed section. The previous section was only related to training the models (with early stopping).

`pred_probs` and `features` are both computed in this section.

File: datalab/README.md [hunk 7]
```diff
@@ -0,0 +1,18 @@
+# Training a Swin Transformer model with PyTorch on a subset of Caltech-256 for auditing with Datalab
+This example demonstrates the use of the following module below from cleanlab:
```
Comment (jwmueller) [inline datalab/README.md]:
```suggestion
This example demonstrates cleanlab's Datalab class for auditing an image dataset.
```

File: datalab/README.md [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+
+  - Install dependencies with:
+
+    ```
+    pip install -r requirements-train.txt
+    ```
+
+- [datalab.ipynb](datalab.ipynb) - Runs the dataset through Datalab with the trained model artifacts.
+  - Install dependencies with
```
Comment (jwmueller) [inline datalab/README.md]:
```suggestion
- [datalab.ipynb](datalab.ipynb) - Audits the dataset using Datalab applied to outputs from the trained model.
  - Install dependencies with
```

File: datalab/README.md [hunk 9]
```diff
@@ -0,0 +1,18 @@
+# Training a Swin Transformer model with PyTorch on a subset of Caltech-256 for auditing with Datalab
+This example demonstrates the use of the following module below from cleanlab:
+
+It is split into two notebooks:
+- [train_image_classifier.ipynb](train_image_classifier.ipynb) - Trains a Swin Transformer model on a subset of Caltech-256
```
Comment (jwmueller) [inline datalab/README.md]:
```suggestion
- [train_image_classifier.ipynb](train_image_classifier.ipynb) - Trains a Swin Transformer classifier model on a subset of Caltech-256
```

File: datalab/README.md [hunk 10]
```diff
@@ -0,0 +1,18 @@
+# Training a Swin Transformer model with PyTorch on a subset of Caltech-256 for auditing with Datalab
```
Comment (jwmueller) [inline datalab/README.md]:
```suggestion
# Detecting issues in an image classification dataset (Caltech-256) with Datalab and a Swin Transformer model.
```

File: datalab/requirements-datalab.txt [hunk 11]
```diff
@@ -0,0 +1,2 @@
+git+https://github.com/elisno/cleanlab.git@datalab#egg=cleanlab[datalab]
```
Comment (jwmueller) [inline datalab/requirements-datalab.txt]: don't forget to update this when next version of cleanlab is released that contains Datalab

File: datalab/requirements-train.txt [hunk 12]
```diff
@@ -0,0 +1,9 @@
+numpy
+opencv-python
+pandas
+scikit-learn
+timm
+tqdm
+--extra-index-url https://download.pytorch.org/whl/cu116
```
Comment (jwmueller) [inline datalab/requirements-train.txt]: is this really something to specify for an arbitrary user?

File: datalab/README.md [hunk 13]
```diff
@@ -0,0 +1,18 @@
+# Detecting issues in an image classification dataset (Caltech-256) with Datalab and a Swin Transformer model.
```
Comment (jwmueller) [inline datalab/README.md:1]: Add a reference to this example in the table of contents from the main readme in examples/

File: datalab/README.md [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+This example demonstrates cleanlab's Datalab class for auditing an image dataset.
```
Comment (jwmueller) [inline datalab/README.md]:
```suggestion
# Detecting issues in an image classification dataset (Caltech-256) with Datalab and a Swin Transformer model.

This example demonstrates cleanlab's Datalab class for auditing an image dataset.
```

File: README.md [hunk 15]
```diff
@@ -8,6 +8,7 @@ To quickly learn how to run cleanlab on your own data, first check out the [quic
 
 |     | Example                                                                                        | Description                                                                                                                                                                                                                                                                  |
 | --- | ---------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| 0   | [datalab](datalab_image_classification)                              | Use Datalab to detect issues in (a subset of) the Caltech-256 image classification dataset.                        |
```
Comment (jwmueller) [inline README.md:11]: TODO: @elisno 

Please rename the entire folder where all these files live to be: `datalab_image_classification` as I've specified in this table of contents link here.

File: datalab/requirements-train.txt (hunk 12 above)
Comment (elisno) [reply datalab/requirements-train.txt]: Moved this to the readme as a pip install instruction like:

```
python -m pip install --extra-index-url http://my.package.repo/simple SomePackage
```

**Review by huiwengoh — APPROVED** [review]
One small question/comment: should we pin some of the dependency versions in `requirements.txt` ?

Otherwise LGTM!

File: datalab/requirements-datalab.txt (hunk 11 above)
Comment (elisno) [reply datalab/requirements-datalab.txt]: Switching to cleanlab/cleanlab repo for now, but keeping this reminder for release.

File: datalab_image_classification/README.md [hunk 16]
```diff
@@ -0,0 +1,19 @@
+# Detecting issues in an image classification dataset (Caltech-256) with Datalab and a Swin Transformer model.
+
+This example demonstrates cleanlab's Datalab class for auditing an image dataset.
+
+It is split into two notebooks:
```
Comment (jwmueller) [inline datalab_image_classification/README.md]:
```suggestion
# Detecting issues in an image classification dataset (Caltech-256) with Datalab

This example uses cleanlab's Datalab class to audit an image dataset. Here we run Datalab with a Swin Transformer model trained for classification.

There are two notebooks:
```

**Review by jwmueller — APPROVED** [review]
LGTM!

**Pull request merged**